tests/migrations/test_executor.py

```diff
@@ -649,6 +653,24 @@ def test_migrate_marks_replacement_applied_even_if_it_did_nothing(self):
             recorder.applied_migrations(),
         )
 
+    @skipIfDBFeature('can_rollback_ddl')
+    @override_settings(MIGRATION_MODULES={"migrations": "migrations.test_migrations_recorded_atomically"})
+    def test_migrations_applied_and_recorded_atomically(self):
+        """
+        Migrations will be applied and recorded atomically.
+        """
+        executor = MigrationExecutor(connection)
+        with mock.patch('django.db.migrations.executor.MigrationExecutor.record_migration') as record_migration:
+            record_migration.side_effect = RuntimeError("Recording migration failed")
+
+            with self.assertRaisesMessage(RuntimeError, "Recording migration failed"):
+                # This will call `apply_migration` which
+                # in-turn calls `record_migration`
+                executor.migrate([("migrations", "0001_initial")], fake=False)
+        # Ensure that migration record is not there
+        migration_recorder = MigrationRecorder(connection)
+        self.assertFalse(migration_recorder.migration_qs.filter(app="migrations", name="0001_initial").exists())
```

Comment:
The important assertions here is that the operation changes were rolledback as well. You want to make sure to call `self.assertTableNotExists('migrations_foo')`.

Reply:
Done with this!